File: clients/javascript/model/action-type.js
/**
 * Type of action a unit is currently performing
 */
class ActionType {
    constructor(name, tag) {
        this.name = name;
        this.tag = tag;
    }

    /**
     * Picking up or dropping loot
     */
    static LOOTING = new ActionType("LOOTING", 0);
    /**
     * Using a shield potion
     */
    static USE_SHIELD_POTION = new ActionType("USE_SHIELD_POTION", 1);

    /**
     * Read ActionType from input stream
     */
    static async readFrom(stream) {
        const tag = await stream.readInt();
        if (tag == ActionType.LOOTING.tag) {
            return ActionType.LOOTING;
        }
        if (tag == ActionType.USE_SHIELD_POTION.tag) {
            return ActionType.USE_SHIELD_POTION;
        }
        throw new Error("Unexpected tag value");
    }

    /**
     * Write ActionType to output stream
     */
    async writeTo(stream) {
        await stream.writeInt(this.tag);
    }

    [Symbol.for('nodejs.util.inspect.custom')]() {
        return this.name;
    }
}

module.exports = ActionType;

File: clients/rust/src/model/action_type.rs
use super::*;

/// Type of action a unit is currently performing
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum ActionType {
    /// Picking up or dropping loot
    Looting,
    /// Using a shield potion
    UseShieldPotion,
}

impl trans::Trans for ActionType {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let tag: i32 = match self {
            Self::Looting => 0,
            Self::UseShieldPotion => 1,
        };
        trans::Trans::write_to(&tag, writer)
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => Ok(Self::Looting),
            1 => Ok(Self::UseShieldPotion),
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}